pallets/liquid-staking/src/types.rs

```diff
@@ -49,59 +57,105 @@ impl<Balance: BalanceT + FixedPointOperand> MatchingLedger<Balance> {
     }
 
-    pub fn update_total_stake_amount(
-        &mut self,
-        amount: Balance,
-        kind: ArithmeticKind,
-    ) -> DispatchResult {
-        use ArithmeticKind::*;
-        match kind {
-            Addition => {
-                self.total_stake_amount = self
-                    .total_stake_amount
-                    .checked_add(&amount)
-                    .ok_or(ArithmeticError::Overflow)?;
-            }
-            Subtraction => {
-                self.total_stake_amount = self
-                    .total_stake_amount
-                    .checked_sub(&amount)
-                    .ok_or(ArithmeticError::Underflow)?;
-            }
-        }
+    pub fn add_stake_amount(&mut self, amount: Balance) -> DispatchResult {
+        self.total_stake_amount.free = self
+            .total_stake_amount
+            .free
+            .checked_add(&amount)
+            .ok_or(ArithmeticError::Overflow)?;
         Ok(())
     }
 
-    pub fn update_total_unstake_amount(
-        &mut self,
-        amount: Balance,
-        kind: ArithmeticKind,
-    ) -> DispatchResult {
-        use ArithmeticKind::*;
-        match kind {
-            Addition => {
-                self.total_unstake_amount = self
-                    .total_unstake_amount
-                    .checked_add(&amount)
-                    .ok_or(ArithmeticError::Overflow)?;
-            }
-            Subtraction => {
-                self.total_unstake_amount = self
-                    .total_unstake_amount
-                    .checked_sub(&amount)
-                    .ok_or(ArithmeticError::Underflow)?;
-            }
-        }
+    pub fn sub_stake_amount(&mut self, amount: Balance) -> DispatchResult {
+        self.total_stake_amount.free = self
+            .total_stake_amount
+            .free
+            .checked_sub(&amount)
+            .ok_or(ArithmeticError::Underflow)?;
+        Ok(())
+    }
+
+    pub fn add_unstake_amount(&mut self, amount: Balance) -> DispatchResult {
+        self.total_unstake_amount.free = self
+            .total_unstake_amount
+            .free
+            .checked_add(&amount)
+            .ok_or(ArithmeticError::Overflow)?;
+        Ok(())
+    }
+
+    pub fn sub_unstake_amount(&mut self, amount: Balance) -> DispatchResult {
+        self.total_unstake_amount.free = self
+            .total_unstake_amount
+            .free
+            .checked_sub(&amount)
+            .ok_or(ArithmeticError::Underflow)?;
+        Ok(())
+    }
+
+    pub fn set_stake_amount_lock(&mut self, amount: Balance) -> DispatchResult {
+        self.total_stake_amount.free = self
+            .total_stake_amount
+            .free
+            .checked_sub(&amount)
+            .ok_or(ArithmeticError::Underflow)?;
+        self.total_stake_amount.reserved = self
+            .total_stake_amount
+            .reserved
+            .checked_add(&amount)
+            .ok_or(ArithmeticError::Overflow)?;
+        Ok(())
+    }
```

Comment:
Maybe we can use two fields: `total_amount` and `reserve_amount`, then `free_amount = total_amount - reserve_amount`